### PR TITLE
Introduce PackageLatestVersionFinder for `npm_and_yarn` behind feature flag Part-1

### PR DESCRIPTION
**Pull request by kbukum1**
### What are you trying to accomplish?

Introduce a subclass of `PackageLatestVersionFinder` under `npm_and_yarn` and integrate it behind a feature flag: `enable_cooldown_for_npm_and_yarn`.

The goal is to align `npm_and_yarn` with the shared `PackageDetailsFetcher` + `PackageLatestVersionFinder` pattern while maintaining backward compatibility.

The new implementation will be used only when the feature flag is enabled, so we can monitor production stability and gradually roll out improvements.

We are intentionally preserving the original `LatestVersionFinder` to support fallback and phased migration.

### Anything you want to highlight for special attention from reviewers?

- This is a non-breaking change; the new logic only activates if the `enable_cooldown_for_npm_and_yarn` experiment is enabled.
- Both implementations share the same business logic for version selection and fallback behavior, including support for dist-tags.
- The new subclass architecture makes it easier to extend and test going forward.

Once we've validated the stability in production, we'll deprecate and remove the old `LatestVersionFinder` class in a future cleanup PR.

### How will you know you've accomplished your goal?

- Tests continue to pass under both implementations.
- In production, enabling the feature flag should activate the new path for specific jobs without changing behavior.
- We can observe differences in logs, debug edge cases more easily, and progressively shift more traffic to the new implementation.

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package/package_details_fetcher.rb [hunk 1]
```diff
@@ -84,8 +89,6 @@ def parse_versions(npm_data)
 
             deprecated = details["deprecated"]
 
-            puts "version: #{version}, #{latest_version}"
-
```
Comment (kbukum1) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package/package_details_fetcher.rb:88]: **Review Tip**: Previously we created package release fetcher but we did forget to remove the `puts`

- Ref: https://github.com/dependabot/dependabot-core/pull/11864

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package/registry_finder.rb [hunk 2]
```diff
@@ -89,6 +89,13 @@ def registry_from_rc(dependency_name)
           explicit_registry_from_rc(dependency_name) || global_registry
         end
 
+        sig { returns(T::Boolean) }
+        def custom_registry?
+          return false if CENTRAL_REGISTRIES.include?(registry_url)
+
+          !(registry_url || "").match?(/registry\.npmjs\.(org|com)/)
+        end
```
Comment (kbukum1) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package/registry_finder.rb:97]: **Review Tip**: Helps determine if the package is using a custom registry.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker.rb [hunk 3]
```diff
@@ -51,7 +51,9 @@ def initialize(dependency:, dependency_files:, credentials:, # rubocop:disable M
         @latest_version_for_git_dependency = T.let(nil, T.nilable(T.any(String, Gem::Version)))
         @latest_released_version = T.let(nil, T.nilable(Gem::Version))
         @latest_version_details = T.let(nil, T.nilable(T::Hash[Symbol, T.untyped]))
-        @latest_version_finder = T.let(nil, T.nilable(LatestVersionFinder))
+        @latest_version_finder = T.let(
+          nil, T.nilable(T.any(LatestVersionFinder, PackageLatestVersionFinder))
+        )
         @version_resolver = T.let(nil, T.nilable(VersionResolver))
```
Comment (kbukum1) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker.rb:57]: **Review Tip**: We're introducing `PackageLatestVersionFinder` in the update checker behind a feature flag. `LatestVersionFinder` should continue to work without any changes.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker.rb [hunk 4]
```diff
@@ -402,17 +404,34 @@ def latest_version_details
...
+            LatestVersionFinder.new(
+              dependency: dependency,
+              credentials: credentials,
+              dependency_files: dependency_files,
+              ignored_versions: ignored_versions,
+              raise_on_ignored: raise_on_ignored,
+              security_advisories: security_advisories
+            )
+          end
+      end
```
Comment (kbukum1) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker.rb:430]: **Review Tip**: We're introducing `PackageLatestVersionFinder` in the update checker behind a feature flag. `LatestVersionFinder` should continue to work without any changes.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb [hunk 5]
```diff
@@ -204,27 +213,49 @@ def dependency_updates_from_full_unlock
...
+              LatestVersionFinder.new(
+                dependency: dep,
+                credentials: credentials,
+                dependency_files: dependency_files,
+                ignored_versions: [],
+                security_advisories: [],
+                raise_on_ignored: raise_on_ignored
+              )
+            end
+        end
```
Comment (kbukum1) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb:254]: **Review Tip**: We're introducing `PackageLatestVersionFinder` in the update checker behind a feature flag. `LatestVersionFinder` should continue to work without any changes.

File: common/lib/dependabot/package/package_latest_version_finder.rb [hunk 6]
```diff
@@ -130,10 +130,15 @@ def fetch_latest_version(language_version:)
         versions = filter_unsupported_versions(version_hashes, language_version)
         versions = filter_prerelease_versions(versions)
         versions = filter_ignored_versions(versions)
-
+        versions = apply_post_fetch_latest_versions_filter(versions)
```
Comment (kbukum1) [inline common/lib/dependabot/package/package_latest_version_finder.rb:133]: **Review Tip**: Subclasses can override `apply_post_fetch_latest_versions_filter` method to apply additional filters after fetch_latest_version filterings. This is applied because `npm_and_yarn` has additional checks for yanked that requires to check through api client.

File: common/lib/dependabot/package/package_latest_version_finder.rb [hunk 7]
```diff
@@ -130,10 +130,15 @@ def fetch_latest_version(language_version:)
...
         versions.max
       end
 
+      sig { params(versions: T::Array[Dependabot::Version]).returns(T::Array[Dependabot::Version]) }
+      def apply_post_fetch_latest_versions_filter(versions)
+        versions
+      end
```
Comment (kbukum1) [inline common/lib/dependabot/package/package_latest_version_finder.rb:140]: **Review**: If subclass doesn't override this method it will return the given versions without any change.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package/package_details_fetcher.rb [hunk 8]
```diff
@@ -80,12 +147,10 @@ def parse_versions(npm_data)
           versions_data.filter_map do |version, details|
             next unless Dependabot::NpmAndYarn::Version.correct?(version)
 
-            package_type = details.dig("repository", "type")
+            package_type = infer_package_type(details)
 
             deprecated = details["deprecated"]
 
-            puts "version: #{version}, #{latest_version}"
```
Comment (kbukum1) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package/package_details_fetcher.rb:87]: **Review Tip**: Removed forgotten `puts`.

**Review by sachin-sandhu — APPROVED** [review]

**Review by sachin-sandhu — APPROVED** [review]
